generic-ec-zkp/src/dlog_eq.rs

```diff
@@ -0,0 +1,277 @@
+        if lhs != rhs {
+            return Err(InvalidProof);
+        }
+
+        Ok(())
+    }
+}
+
+/// Functions for non interactive variant of the proof. See usage example in
+/// [`dlog_eq`](crate::dlog_eq)
+///
+/// Compared to naive protocol produced by Fiat-Shamir heuristic, this is
+/// optimized to send less data in proof. It's based on `PrEq` functionality in
+/// this paper: <https://eprint.iacr.org/2020/096>
+#[cfg(feature = "udigest")]
+pub mod non_interactive {
+    use generic_ec::{Curve, Point, Scalar, SecretScalar};
+
+    const TAG: &str = "bls-style-digest.zkp";
+
+    pub use super::Data;
+    pub use super::InvalidProof;
+
+    /// Proof data, sent by prover to the verifier
+    #[derive(Debug, Clone, Copy)]
+    #[cfg_attr(
+        feature = "serde",
+        derive(serde::Serialize, serde::Deserialize),
+        serde(bound = "")
+    )]
+    pub struct Proof<E: generic_ec::Curve> {
+        /// Deterministic challenge
+        pub ch: generic_ec::Scalar<E>,
+        /// Proof itself
+        pub res: generic_ec::Scalar<E>,
+    }
+
+    /// Create a proof deterministically for given data
+    ///
+    /// - `shared_state` - shared data not known to this proof, used to protect
+    ///   from replay attacks
+    /// - `share` - `x`, secret value of discrete logarithm
+    /// - `data` - data to compute the proof for
+    /// - `r` - random nonce. Can be produced by `Scalar::random(&mut rng)`
+    pub fn prove<D: digest::Digest, E: Curve>(
+        shared_state: &impl udigest::Digestable,
+        share: &SecretScalar<E>,
+        data: Data<E>,
+        r: Scalar<E>,
+    ) -> Proof<E> {
+        let com1 = Point::generator() * r;
+        let com2 = data.base * r;
+
+        let seed = udigest::inline_struct!(TAG {
+            shared_state,
+            data,
+            com1,
+            com2,
+        });
+        let mut rng = rand_hash::HashRng::<D, _>::from_seed(seed);
+        let ch = Scalar::random(&mut rng);
+
+        let res = r + share * ch;
+        Proof { ch, res }
+    }
+
+    /// Verify the validity of the ZK proof
+    ///
+    /// - `shared_state` - shared data not known to this proof, used to protect
+    ///   from replay attacks
+    /// - `data` - data for which the proof was computed
+    /// - `proof` - produced by the Prover in the second round for this data and
+    ///   shared_state
+    pub fn verify<D: digest::Digest, E: Curve>(
+        shared_state: &impl udigest::Digestable,
+        data: Data<E>,
+        proof: Proof<E>,
+    ) -> Result<(), InvalidProof> {
+        let com1 = Point::generator() * proof.res - data.pub_share * proof.ch;
+        let com2 = data.base * proof.res - data.exp * proof.ch;
+
+        let seed = udigest::inline_struct!(TAG {
+            shared_state,
+            data,
+            com1,
+            com2,
+        });
+        let mut rng = rand_hash::HashRng::<D, _>::from_seed(seed);
+        let ch = Scalar::random(&mut rng);
```

Comment:
You probably want to use [`Scalar::from_hash`](https://docs.rs/generic-ec/latest/generic_ec/struct.Scalar.html#method.from_hash) instead